src/Infrastructure/Service/Precondition/NoLinksPointOutsideTheCodebase.php

```diff
@@ -0,0 +1,135 @@
+<?php declare(strict_types=1);
+
+namespace PhpTuf\ComposerStager\Infrastructure\Service\Precondition;
+
+use PhpTuf\ComposerStager\Domain\Exception\InvalidArgumentException;
+use PhpTuf\ComposerStager\Domain\Exception\IOException;
+use PhpTuf\ComposerStager\Domain\Service\Filesystem\FilesystemInterface;
+use PhpTuf\ComposerStager\Domain\Service\Precondition\NoLinksPointOutsideTheCodebaseInterface;
+use PhpTuf\ComposerStager\Domain\Value\Path\PathInterface;
+use PhpTuf\ComposerStager\Domain\Value\PathList\PathListInterface;
+use PhpTuf\ComposerStager\Infrastructure\Factory\Path\PathFactoryInterface;
+use PhpTuf\ComposerStager\Infrastructure\Service\Finder\RecursiveFileFinderInterface;
+use PhpTuf\ComposerStager\Infrastructure\Value\PathList\PathList;
+
+final class NoLinksPointOutsideTheCodebase extends AbstractPrecondition implements
+    NoLinksPointOutsideTheCodebaseInterface
+{
+    private RecursiveFileFinderInterface $fileFinder;
+
+    private FilesystemInterface $filesystem;
+
+    private PathFactoryInterface $pathFactory;
+
+    private string $unfulfilledStatusMessage = <<<'EOF'
+The %s directory at "%s" contains links that point outside the codebase, which is not supported. The first one is "%s".
+EOF;
```

Comment:
🤔 AFAICT this means that this allows more than just _intra-package_ links: it also allows _inter-package_ links, and even links from within a package to e.g. an uploaded/generated file (if the directory containing user-uploaded files or image derivatives such as thumbnails).

I see that https://github.com/php-tuf/composer-stager/issues/58 also doesn't distinguish between those cases, but your comment at https://www.drupal.org/project/automatic_updates/issues/3319507#comment-14855463 _did_:
> 4. Completely forget about links that cross package boundaries.

Reply:
@wimleers I'm going to delete this comment, because it's based on misunderstanding and inaccuracies. I'm afraid it just clutters the discussion here with misleading chatter. I will say in summary that your concerns were already accounted for. Thanks.